pkg/scene/widget_section_test.go

```diff
@@ -0,0 +1,142 @@
+	wid := id.NewWidgetID()
+	wid2 := id.NewWidgetID()
+	wid3 := id.NewWidgetID()
+
+	ws := NewWidgetSection()
+	ws.top.widgetIds = append(ws.top.widgetIds, wid)
+	ws.middle.widgetIds = append(ws.middle.widgetIds, wid2)
+	ws.bottom.widgetIds = append(ws.bottom.widgetIds, wid3)
+	top := ws.top
+	mid := ws.middle
+	bot := ws.bottom
+
+	testCases := []struct {
+		Name     string
+		Input    id.WidgetID
+		WS       *WidgetSection
+		Expected *WidgetArea
+	}{
+		{
+			Name:     "Find the location (top) of a widgetID and return the WidgetArea",
+			Input:    wid,
+			WS:       ws,
+			Expected: &top,
+		},
+		{
+			Name:     "Find the location (middle) of a widgetID and return the WidgetArea",
+			Input:    wid2,
+			WS:       ws,
+			Expected: &mid,
+		},
+		{
+			Name:     "Find the location (bottom) of a widgetID and return the WidgetArea",
+			Input:    wid3,
+			WS:       ws,
+			Expected: &bot,
+		},
+		{
+			Name:     "Return nil if no widget section",
+			Input:    wid3,
+			WS:       nil,
+			Expected: nil,
+		},
+	}
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.Name, func(tt *testing.T) {
+			tt.Parallel()
+			_, res := tc.WS.Find(tc.Input)
+			assert.Equal(tt, tc.Expected, res)
+		})
+	}
+}
+
+func TestWidgetSection_Remove(t *testing.T) {
+	wid := id.NewWidgetID()
+	ws := NewWidgetSection()
+	ws.top.widgetIds = append(ws.top.widgetIds, wid)
+
+	testCases := []struct {
+		Name         string
+		Input        id.WidgetID
+		WS, Expected *WidgetSection
+	}{
+		{
+			Name:     "Remove a widget from widget section",
+			Input:    wid,
+			WS:       ws,
+			Expected: &WidgetSection{top: WidgetArea{widgetIds: []id.WidgetID{}}},
+		},
+		{
+			Name:     "Return nil if no widget section",
+			Input:    wid,
+			WS:       nil,
+			Expected: nil,
+		},
+	}
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.Name, func(tt *testing.T) {
+			tt.Parallel()
+			tc.WS.Remove(tc.Input)
+			assert.Equal(tt, tc.Expected, tc.WS)
+		})
+	}
+}
+
+func TestWidgetSection_Area(t *testing.T) {
+	wid := id.NewWidgetID()
+
+	ws := NewWidgetSection()
+	ws.top.widgetIds = append(ws.top.widgetIds, wid)
+
+	testCases := []struct {
+		Name     string
+		Input    string
+		WS       *WidgetSection
+		Expected *WidgetArea
+	}{
+		{
+			Name:     "From a Widget Section return top Widget Area",
+			Input:    "top",
+			WS:       ws,
+			Expected: &ws.top,
+		},
+		{
+			Name:     "From a Widget Section return middle Widget Area",
+			Input:    "middle",
+			WS:       ws,
+			Expected: &ws.middle,
+		},
+		{
+			Name:     "From a Widget Section return bottom Widget Area",
+			Input:    "bottom",
+			WS:       ws,
+			Expected: &ws.bottom,
+		},
+		{
+			Name:     "Return nil when can't find Widget Area",
+			Input:    "topMiddleBottom",
+			WS:       ws,
+			Expected: nil,
+		},
+	}
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.Name, func(tt *testing.T) {
+			tt.Parallel()
+			res := tc.WS.Area(tc.Input)
+			assert.Equal(tt, tc.Expected, res)
```

Comment:
Just to make sure, the shared approach is okay sometimes right? Because quite a few tests need to include one where the method's receiver doesn't exist, which I can't think of how that works if I move the WZ, WA, etc into the for loop.
In other tests (for Widgets, widget system, etc) we don't do as you are suggesting here, but share them and those also have receiver pointers that check if they are nil or not.
Places were data races have occured I removed the `tt.Parallel()` which for obvious reasons avoided the data races. 
Please let me know if you would still like me to do as you suggested here. And if so, please let me know if it is fine I ignore CodeCov's warning that I am not testing the nil receiver.

Reply:
Using the same object in unit tests is not a good idea because it weakens the independence of each test case. Since each test case is less independent, it may not be possible to run them in parallel and data race may occur.

It is easy to make test cases independent. Just confine the creation of structs that may be mutated in each test inside the test case.

Not good:

```go
a := NewHoge()
b := "hoge"

for tc := range testCases {
	tc := tc
	t.Run(tc.Name, func(tt *testing.T) {
		tt.Parallel()

		a.Update() // This may cause data race
        // Using b here is ok because b is used as read-only and not mutated
	})
}
```

Good:

```go
b := "hoge"

for tc := range testCases {
	tc := tc
	t.Run(tc.Name, func(tt *testing.T) {
		tt.Parallel()

		a := NewHoge()
		a.Update() // no problem
	})
}
```

Reply:
Unit tests written by others can be left as they are, so for the newly written unit tests in this PR, I would like you to modify them to make the code as good as possible.